build_runner/lib/src/package_graph/apply_builders.dart

```diff
@@ -5,59 +5,68 @@
     (p) => packages.contains(p.name);
 
+/// Indcates that [builder] should be run against all packages matching
+/// [filter].
+///
+/// If the builder should only run on a subset of files within a package pass
+/// globs to [inputs] or [excludes];
+BuilderApplication apply(Builder builder, PackageFilter filter,
+        {List<String> inputs, List<String> excludes, bool isOptional}) =>
+    new BuilderApplication(builder, filter,
+        inputs: inputs, excludes: excludes, isOptional: isOptional);
+
 /// A description of which packages need a given [Builder] applied.
 class BuilderApplication {
   final Builder builder;
 
   /// Determines whether a given package needs [builder] applied.
-  final _PackageFilter _filter;
-
-  /// Apply [builder] to all packages in the dependency graph.
-  const BuilderApplication.allPackages(this.builder) : _filter = _alwaysTrue;
-
-  /// Apply [builder] to the packages which have a direct dependency on the
-  /// package named [dependency].
-  BuilderApplication.dependentsOf(this.builder, String dependency)
-      : _filter = _hasDependency(dependency);
+  final PackageFilter filter;
 
-  /// Apply [builder] to a single package in the graph.
-  BuilderApplication.singlePackage(this.builder, String package)
-      : _filter = _isPackage(package);
+  final List<String> inputs;
+  final List<String> excludes;
+  final bool isOptional;
 
-  /// Apply [builder] to a set of explicitly listed packages.
-  BuilderApplication.onPackages(this.builder, Set<String> packages)
-      : _filter = _isInSet(packages);
+  const BuilderApplication(this.builder, this.filter,
+      {this.inputs, this.excludes, this.isOptional});
 }
 
 /// Creates a [BuildAction] to apply each builder in [builders] to each package
-/// in [packages] such that all builders are run for earlier subcomponents
-/// before moving on to later packages.
+/// in [packageGraph] such that all builders are run for dependencies before
+/// moving on to later packages.
 ///
-/// For example, given the builders `[b1, b2]` and package components
-/// `[[p1], [p2, p3], [p4]]` where `p2` and `p3` represent a dependenency cycle
-/// which depends on `p1` this will return actions
-/// `[b1p1, b2p1, b1p2, b1p3, b2p2, b2p3, b1p4, b2p4]`.
+/// When there is a package cycle the builders are applied to each packages
+/// within the cycle before moving on to packages that depend on any package
+/// within the cycle.
 ///
 /// Builders may be filtered, for instance to run only on package which have a
-/// dependency on some other package.
-List<BuildAction> applyBuilders(Iterable<Iterable<PackageNode>> packages,
-        Iterable<BuilderApplication> builders) =>
-    packages
-        .expand((cycle) => builders.expand((b) => cycle
-            .where(b._filter)
-            .map((p) => new BuildAction(b.builder, p.name))))
+/// dependency on some other package by choosing the appropriate
+/// [BuilderApplication].
+List<BuildAction> applyBuilders(
```

Comment:
This name doesn't make so much sense any more imo, maybe `createBuildActions` or something like that? The 2nd arg should probably be renamed `builderApplications` as well.

The main thing I think is confusing is the difference between `apply` and `applyBuilders`, it isn't clear from the names how they should be used and in what order.

Reply:
`createBuildActions` SGTM